onnxruntime/test/providers/nnapi/nnapi_basic_test.cc

```diff
@@ -388,26 +406,133 @@ TEST(NnapiExecutionProviderTest, TestQDQReshape) {
 TEST(NnapiExecutionProviderTest, TestQDQSoftMax) {
   RunQDQModelTest(BuildQDQSoftMaxTestCase<uint8_t, uint8_t>(
                       {1, 32} /* input_shape */,
-                      static_cast<int64_t>(1) /* axis */),
+                      static_cast<int64_t>(1) /* axis */,
+                      1.f / 256 /* output_scales */,
+                      0 /* output_zp */),
                   "nnapi_qdq_test_graph_softmax",
                   {
                       true /* verify_entire_graph_use_ep */
                   });
 }
 
+// See https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc#L1236
+TEST(NnapiExecutionProviderTest, TestQDQSoftMax_UnsupportedOutputScaleAndZp) {
+  RunQDQModelTest(BuildQDQSoftMaxTestCase<uint8_t, uint8_t>(
+                      {1, 32} /* input_shape */,
+                      static_cast<int64_t>(1) /* axis */,
+                      0.002f /* output_scales */,
+                      1 /* output_zp */),
+                  "nnapi_qdq_test_graph_softmax",
+                  {
+                      true /* verify_entire_graph_use_ep */
+                  },
+                  false /* nnapi_qdq_model_supported */);
+}
+
 TEST(NnapiExecutionProviderTest, TestQDQConcat) {
+  // This is to verify all the inputs have the same scale and zp as input 0
+  // See https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_support_checker.cc#L16176
+  auto check_graph = [](Graph& test_graph) {
```

Comment:
Supposedly at least from what I know for operators: AveragePool, Transpose, Reshape, Binary op, the input&output scales and zero points should match for NNAPI. Haven't added the check for these, (should be the same/similar thing)
Only added check_graph function for concat as it appears to be a bit long process to achieve the scale and zp. open to suggestions.

Reply:
could probably put the extraction of scale and zp values into a test helper function to make it more concise

Reply:
actually, do we need to check this? if I understand correctly, this is just verifying the graph built from other test code.

Reply:
yes you are right. I am not so sure about this either, just as I saw in the `add concat op` pr, there are comments about passing in lambda function to check graph/model. so I added these. probably meant for other check?

Reply:
copying from the linked comment:

> So far we don't have tests for that, but this can be easily added

> We probably need to add a lambda function to verify the model in the test instead of just verify the number of nodes taken by the EP (this can be the default verification)

@gwang-msft What model verification is this referring to?

Reply:
For now we only use the number of nodes taken by the EP, (for example, none, some, all), there is no graph verification after the partitioning, we could probably add some optional verification function to verify the exact nodes taken by the EP (the function can be defaulted to verify number of nodes like what we are doing now)

The verify graph function can be similar to the `check_graph` in the `qdq_transformer_test`